arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java

```diff
@@ -407,26 +412,36 @@ public void setBatchSize(int batchSize) {
   }
 
   private static final class PositionVectorReader extends VectorizedArrowReader {
+    private final Field arrowField = ArrowSchemaUtil.convert(MetadataColumns.ROW_POSITION);
+    private final BufferAllocator bufferAllocator = ArrowAllocation.rootAllocator();
+    private final boolean setArrowValidityVector;
     private long rowStart;
+    private int batchSize;
+    private FieldVector vec;
     private NullabilityHolder nulls;
 
+    PositionVectorReader(boolean setArrowValidityVector) {
+      this.setArrowValidityVector = setArrowValidityVector;
+    }
+
     @Override
     public VectorHolder read(VectorHolder reuse, int numValsToRead) {
-      Field arrowField = ArrowSchemaUtil.convert(MetadataColumns.ROW_POSITION);
-      FieldVector vec = arrowField.createVector(ArrowAllocation.rootAllocator());
-
-      if (reuse != null) {
+      if (reuse == null) {
+        this.vec = newVector();
+        this.nulls = new NullabilityHolder(batchSize);
+      } else {
         vec.setValueCount(0);
         nulls.reset();
-      } else {
-        ((BigIntVector) vec).allocateNew(numValsToRead);
-        for (int i = 0; i < numValsToRead; i += 1) {
-          vec.getDataBuffer().setLong(i * Long.BYTES, rowStart + i);
-        }
-        for (int i = 0; i < numValsToRead; i += 1) {
-          BitVectorHelper.setBit(vec.getValidityBuffer(), i);
+      }
+
+      ArrowBuf dataBuffer = vec.getDataBuffer();
+      ArrowBuf validityBuffer = vec.getValidityBuffer();
+
+      for (int i = 0; i < numValsToRead; i += 1) {
```

Comment:
I switched to a single for loop and setting the Arrow validity vector only if required.

Reply:
I think this should move back to two loops. The point of two is to avoid `if` statements in a loop and to maximize the chances that the memory regions that are used are in cache. High-level control flow and narrow memory access are how we get the benefits of vectorization.

Reply:
We use exactly the same pattern in a few other places. I kind of hoped the complier would be smart enough to rewrite this efficiently as the condition is static but I agree with you. I'll update.

Reply:
We do? We should definitely avoid branching inside these loops!